### PR TITLE
Search cancellation after configured number of seconds

### PR DESCRIPTION
**Pull request by luk-kaminski**
## Description
Introduces new config option in `org.graylog2.indexer.searches.SearchesClusterConfig` Mongo `cluster_config` document, called `cancel_after_seconds`.
Introduces the same param in `org.graylog.plugins.views.search.rest.ExecutionStats`, which can be used to override global setting.
Because of that, in the search execution area there is a possibility to execute a search with an option to cancel automatically if certain time limit is breached.
There is also a possibility to define default value of `cancel_after_seconds`, used automatically in chosen areas of the system(currently SearchResource).

OS version is fine-grained - each search type can be cancelled independently because of the time limit. One can receive results for some widgets.
ES version is coarse-grained - either the whole search succeeds or is cancelled.
/nocl

## Assumptions
`cancel_after_seconds` from config is not applied to reports - they have to finish even if they take long time to execute.
`cancel_after_seconds` from config is not applied to events-related searches.
`cancel_after_seconds` from config is not applied to scripting API calls.

## Screenshots (if appropriate):
Partial retrieval in OS: widgets for long running search types have been cancelled, faster widgets are rendered.
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-web-interface/src/components/configurations/SearchesConfig.tsx [hunk 1]
```diff
@@ -253,6 +273,14 @@ const SearchesConfig = () => {
         </dd>
       </dl>
 
+      <dl className="deflist">
+        <dt>Cancellation timeout</dt>
+        <dd>{cancellationTimeout(viewConfig)}</dd>
+        <dd>The time in seconds per widget after which search execution will be canceled automatically.
+          That minimizes amount of executed searches and improves performance.
```
Comment (linuspahl) [inline graylog2-web-interface/src/components/configurations/SearchesConfig.tsx:280]: Maybe let's say `This minimizes the amount of executed searches and improves performance`?

**Comment by bernd** [comment]
Preface: I might miss some context since I can't find a linked pitch or GitHub issue in this PR.

A global search timeout is a pretty big hammer that is pretty hard to configure when it shouldn't be annoying for users. However, I can see the use case for a global search timeout for admins managing OpenSearch load.

I wonder if regular users would use a configurable timeout per search request. As a user of Graylog, I can't see myself using a per-search request timeout. I wouldn't know what timeout value to use.
I don't want to think about how long I want to wait. But I want a "cancel" button if the search takes too long, and I want to stop it to refine the search.

So, from my perspective, we should rather implement a "cancel" button (if not already done) that stops the search task in OpenSearch and skip the per-request timeout. Again, I might miss some context about the decision to implement this feature.

File: graylog2-web-interface/src/components/configurations/SearchesConfig.tsx [hunk 2]
```diff
@@ -136,6 +136,24 @@ const SearchesConfig = () => {
     setFormConfig({ ...formConfig, query_time_range_limit: queryTimeRangeLimit });
   };
 
+  const isCancelAfterSecondsEnabled = (config: { cancel_after_seconds: number }) => !!config?.cancel_after_seconds;
+
+  const onCancelAfterSecondsChanged = ({ target: { value } }) => {
```
Comment (linuspahl) [inline graylog2-web-interface/src/components/configurations/SearchesConfig.tsx:141]: Let's type the function parameters.

**Review by linuspahl — CHANGES_REQUESTED** [review]
When you remove the content of the cancellation timeout input, it makes the input disappear.

query

**Comment by luk-kaminski** [comment]
> Preface: I might miss some context since I can't find a linked pitch or GitHub issue in this PR.
> 
> A global search timeout is a pretty big hammer that is pretty hard to configure when it shouldn't be annoying for users. However, I can see the use case for a global search timeout for admins managing OpenSearch load.
> 
> I wonder if regular users would use a configurable timeout per search request. As a user of Graylog, I can't see myself using a per-search request timeout. I wouldn't know what timeout value to use. I don't want to think about how long I want to wait. But I want a "cancel" button if the search takes too long, and I want to stop it to refine the search.
> 
> So, from my perspective, we should rather implement a "cancel" button (if not already done) that stops the search task in OpenSearch and skip the per-request timeout. Again, I might miss some context about the decision to implement this feature.

Hi Bernd!
When it comes to your comment - a normal user has no way in the UI to configure the cancel timeout.
Only admin can do this in configuration.
A potential to have this timeout used per call is only about manually using our API, outside UI.
When someone could put it in the JSON of ExecutionStats object.

**Comment by bernd** [comment]
> When it comes to your comment - a normal user has no way in the UI to configure the cancel timeout.
> Only admin can do this in configuration.
> A potential to have this timeout used per call is only about manually using our API, outside UI.
> When someone could put it in the JSON of ExecutionStats object.

Ah okay. Thanks for the clarification! :+1:

File: graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java [hunk 3]
```diff
@@ -77,6 +77,8 @@
 import static com.google.common.base.Strings.isNullOrEmpty;
 import static java.util.stream.Collectors.toMap;
 import static java.util.stream.Collectors.toSet;
+import static org.graylog.plugins.views.search.SearchJob.NO_CANCELLATION;
+import static org.graylog.plugins.views.search.SearchJob.NO_CANCELLATION;
```
Comment (janheise) [inline graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java:81]: import doubled?

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/errors/SearchTypeErrorParser.java [hunk 4]
```diff
@@ -43,6 +48,15 @@ public static SearchTypeError parse(Query query, String searchTypeId, Elasticsea
         return new SearchTypeError(query, searchTypeId, ex);
     }
 
+    private static boolean isSearchTypeAbortedError(ElasticsearchException ex) {
+        return ex != null &&
+                (
+                        ex.getMessage().contains("type=task_cancelled_exception")
+                                || (ex.getCause() != null && ex.getCause() instanceof TimeoutException)
```
Comment (janheise) [inline graylog2-server/src/main/java/org/graylog/plugins/views/search/errors/SearchTypeErrorParser.java:55]: Is this always correct, that a TimeoutException as a cause means it got aborted? Could the timeout not point to other problems in the subsystem?

File: graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java (hunk 3 above)
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java]: Well spotted! Thx!

File: graylog2-server/src/main/java/org/graylog/plugins/views/search/errors/SearchTypeErrorParser.java (hunk 4 above)
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog/plugins/views/search/errors/SearchTypeErrorParser.java:55]: Maybe `aborted` is a bad name and I should name it differently? 
As for me it is not that important if it is a cancel or timeout, it is important it did not finish, which I called aborted...

**Review by linuspahl — APPROVED** [review]

File: graylog2-server/src/main/java/org/graylog2/indexer/searches/SearchesClusterConfig.java [hunk 5]
```diff
@@ -115,7 +121,8 @@ public static SearchesClusterConfig create(@JsonProperty("query_time_range_limit
                                                @JsonProperty("analysis_disabled_fields") @Nullable Set<String> analysisDisabledFields,
                                                @JsonProperty("auto_refresh_timerange_options") @Nullable Map<Period, String> autoRefreshTimerangeOptions,
                                                @JsonProperty("default_auto_refresh_option") Period defaultAutoRefreshOption,
-                                               @JsonProperty("quick_access_timerange_presets") @Nullable List<TimerangePreset> quickAccessTimerangePresets
+                                               @JsonProperty("quick_access_timerange_presets") @Nullable List<TimerangePreset> quickAccessTimerangePresets,
+                                               @JsonProperty("cancel_after_interval") @Nullable Integer cancelAfterSeconds
```
Comment (janheise) [inline graylog2-server/src/main/java/org/graylog2/indexer/searches/SearchesClusterConfig.java:125]: Is `cancel_after_interval` correct here? everywhere else it's `cancel_after_seconds`

**Review by janheise — CHANGES_REQUESTED** [review]

**Review by janheise — APPROVED** [review]
Thx. for the last small change. LGTM!

**Pull request merged**